### PR TITLE
bpf: always enforce policy on SYN packets and rewrite closed CT entries with the same tuple

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

    bpf: TCP SYN recycles CT entry if it looks closed
    
    If a SYN matches a CT entry and the entry has seen FIN or RST in both
    directions, it is clearly a new unrelated connection with the same
    tuple and the entry is not clean up yet. Then recycle the entry, report,
    that the flow is new and the entry will be over written by a fresh one.
    
    That also forces the new connection through policy rather than reusing
    the decision made on the old one.

    bpf: enforce policy on SYN packets
    
    In case policy has changed and there is an existing conntrack entry for
    previously allowed connection, a new connection with the same tuple
    would hit the CT entry and would be let through even if the current
    policy does not let it anymore.
    
    This is a corner case, but a client can enforce its source port as our
    named ports FV tests do.

    Unlike NAT lookup, CT lookup does not set the DNAT in state, which is
    fixed only after policy. Therefore we need to set it properly if we want
    to enforce policy correctly.
    
    Adds a test that changes the policy and checks that the policy is
    enforce on a retransmitted SYN - equal to a fresh new SYN hitting an
    existing CT from previous connection.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
BPF mode: improve handling of connections that reuse the same TCP source/dest tuple.  Detect reuse and treat as new connection.
```

### CODE REVIEW

**Comment by tomastigera** [comment]
@caseydavenport this reverets the workaround from https://github.com/projectcalico/calico/pull/5302 while it keeps the options for connectivity checker (which also revived optional descriptions)

File: felix/bpf-gpl/tc.c [hunk 1]
```diff
@@ -263,6 +263,11 @@ static CALI_BPF_INLINE int calico_tc(struct __sk_buff *skb)
 			ctx.state->flags |= CALI_ST_SKIP_FIB;
 		}
 		CALI_DEBUG("CT Hit\n");
+
+		if (ctx.state->ip_proto == IPPROTO_TCP && ct_result_is_syn(ctx.state->ct_result.rc)) {
+			CALI_DEBUG("Forcing policy on SYN\n");
+			goto syn_force_policy;
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:269]: This skips the socket cookie lookup; could that mean that policy is missing some information?

File: felix/bpf-gpl/conntrack.h [hunk 2]
```diff
@@ -788,7 +813,13 @@ static CALI_BPF_INLINE struct calico_ct_result calico_ct_v4_lookup(struct cali_t
 		result.ifindex_fwd = dst_to_src->ifindex;
 	}
 
-	CALI_CT_DEBUG("result: %d\n", result.rc);
+	if (syn) {
+		CALI_CT_DEBUG("packet is SYN\n");
```
Comment (fasaxc) [inline felix/bpf-gpl/conntrack.h:817]: What about SYN-ACK packets, seems like they'd get the new treatment too?

**Review by fasaxc — CHANGES_REQUESTED** [review]
Couple of queries; not sure I fully grok it!

File: felix/bpf-gpl/conntrack.h (hunk 2 above)
Comment (tomastigera) [reply felix/bpf-gpl/conntrack.h:817]: no

`bool syn = tcp_header && tcp_header->syn && !tcp_header->ack;`

File: felix/bpf-gpl/tc.c (hunk 1 above)
Comment (tomastigera) [reply felix/bpf-gpl/tc.c:269]: :thinking: will have a look

Comment (tomastigera) [reply felix/bpf-gpl/tc.c:269]: Yes, we should include that. We should only skip over NAT and RPF because that has been done as part of creating the CT entry that we just hit.

File: felix/bpf-gpl/tc.c [hunk 3]
```diff
@@ -365,11 +371,13 @@ static CALI_BPF_INLINE int calico_tc(struct __sk_buff *skb)
 		ctx.state->nat_dest.port = 0;
 	}
 
+syn_force_policy:
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:374]: Still looks like there are some lines that you jump over that are involved in setting up the state for the policy program.  Don't we need to fill in `nat_dest.*`, `pol_rc` and any flags to match the connection?

In general, there could be pre-DNAT policy and normal policy so I think all the fields would need to be filled in.

Comment (tomastigera) [reply felix/bpf-gpl/tc.c:374]: I think you are right. Things are filled up correctly for the `calico_tc_skb_accepted` but policy uses the `state` and that isn't filled up correctly. I am working on a UT test.

Comment (tomastigera) [reply felix/bpf-gpl/tc.c]: I pushed a fix and a test

File: felix/bpf/ut/nat_test.go [hunk 4]
```diff
@@ -1440,6 +1444,36 @@ func TestNATSYNRetryGoesToSameBackend(t *testing.T) {
 		}
 		Expect(seenOtherIP).To(BeTrue(), "SYNs from varying source ports all went to same backend")
 	})
+
+	// Change back to the
```
Comment (fasaxc) [inline felix/bpf/ut/nat_test.go:1448]: Comment tails off

File: felix/fv/named_ports_test.go [hunk 5]
```diff
@@ -216,17 +215,6 @@ func describeNamedPortTests(testSourcePorts bool, protocol string) {
 		applyAtOthers
 	)
 
-	bpfEnabled := false
```
Comment (fasaxc) [inline felix/fv/named_ports_test.go:219]: What about UDP?  Might still need the clean in that case

**Review by fasaxc — CHANGES_REQUESTED** [review]
One nit and one question on the tests.

File: felix/fv/named_ports_test.go (hunk 5 above)
Comment (tomastigera) [reply felix/fv/named_ports_test.go:219]: how could that differ from what iptables do? There is no sense of a new connection :thinking:

**Review by fasaxc — APPROVED** [review]
